### ISSUE TITLE
Clarifying scope of bindings

### ISSUE THREAD

**Issue by anderslanglands** [vfx-rs/organization#9]
## Introduction

I'd like to propose defining our base "product" of the binding effort as what I'm calling "unsafe+" crates, which I define more clearly below but essentially just means a thin Rusty wrapper over the -sys layer without going full-blown "completely safe" API. The intention is to allow us to focus our efforts and arrive more quickly at a "finished" product that we can release and get into version lockstep with the target libraries, while still providing an ergonomic and flexible API.

So far our stated mission has been generating C, -sys ("unsafe") and -rs ("safe") bindings for all the libraries in [our target list](https://github.com/vfx-rs/organization#tracked-libraries). We've never really defined what exactly "safe" and "unsafe" mean in the context of the bindings - the general assumption has been that the "safe" -rs bindings would present an entirely safe API, with any use of unsafe being hidden behind that API and carefully checked to make sure that users cannot inadvertently violate Rust's safety guarantees. 

Another stated goal is that our Rust APIs should be very close to the C++ API wherever possible, in order not to reinvent the world, and to enable users familiar with the C++ and Python VFX ecosystem to migrate to Rust with as little friction as possible.

Finally, we would like to lock versions with the target libraries where possible. OpenEXR have indicated they're happy to do this, and we expect at least some others to do the same. This is great from a user's point of view when they want to integrate with a specific DCC or VFX Reference Platform version, it's obvious which version of the crate to use. It does however mean we have to be very careful not to break API except at major version updates, which are generally fairly infreuent, particularly for more stable projects like OpenEXR.

## The Problem

These goals are all valuable, but they are ultimately a little bit in conflict with one another. The problem is that C++ APIs are generally inherently unrepresentable in safe Rust. For instance, take the FrameBuffer API from OpenEXR:

```c++
void writeRgba1 (const char fileName[],
                    const Rgba *pixels,
                    int width,
                    int height) {
    RgbaOutputFile file (fileName, width, height, WRITE_RGBA); // 1
    file.setFrameBuffer (pixels, 1, width);                    // 2
    file.writePixels (height);                                 // 3
}
```
The important point here is that `RgbaOutputFile::setFrameBuffer()` stores a reference (pointer) to a block of pixel data, and then reads from it in `RgbaOutputFile::writePixels()`. Between those two calls of course, the user could `delete[] pixels`.

The natural way to express this in safe Rust would be to express `pixels` as a slice and store that slice in the `RgbaOutputFile`. Now we can guarantee that `pixels` is still valid when `writePixels()` is called. 

```rust
struct RgbaOutputFile<'p> {
    inner: *mut sys::Imf_RgbaOutputFile_t,
    pixels: &'p [Rgba],
}
```

Leaving aside the ugliness of the added lifetime annotation and the fact we're now storing our own copy of the pointer which the C++ side also stores, now we've also blocked the user from doing anything else with `pixels` in the meantime. For instance, `pixels` might represent the framebuffer of a progressive renderer that is updating on another thread and which the user wants to periodically write to the file. This would not now be possible as the `RgbaOutputFile` is holding a shared reference to the memory that the renderer wants to write into.

At this stage you're probably thinking of lots of ways around this. For instance, we could wrap `pixels` as a `Arc<Mutex<Vec[Rgba]>>` to allow safe shared access, or we could pass ownership of the pixel data to directly into `RgbaOutputFile` and give back a handle to allow modifying it safely via a closure or some other mechanism.

These and others are all valid solutions to making an unsafe API safe, but they all come with tradeoffs, either in the perfomance overhead of a `Arc<Mutex<>>`, the inflexibility of storing shared references (not to mention lifetime annotation explosion) or the 'weirdness' of closure-based access. Whatever solution we come up with for our safe API might seem great for the use cases we imagine at the time, but you can be sure that there's a workflow that more than one person depends on in C++ that would be blocked by it. 

This is just a trivial example: the general Input/OutputFile API in OpenEXR take multiple Slices. The situation gets *much* worse with libraries like USD, where pretty much every object is essentially a mutable reference to the UsdStage.

The concern here is that attempting to find a "one size fits all" safe API is potentially a huge amount of work, that would require iterating over a long period of time and adjusting as new use cases came to light, if it's even possible at all. This would necessarily lead to breaking API constantly, meaning that we'd either need to stay pre-1.0 for a very long time, or tick the major version up every few weeks.

## The Solution

What I propose instead is that we don't try and go for the holy grail of a completely safe API initially, and instead make the main -rs bindings an "unsafe+" layer. What this means is creating a thin wrapper on top of the -sys crate that transforms it from a FFI binding to an ergonomic Rust binding with proper error handling, Rust types, traits and iterators, using references and slices rather than pointers etc.

For instance, for the RgbaOutputFile above we might have:

```rust
struct RgbaOutputFile(*mut sys::Imf_RgbaOutputFile_t);

impl RgbaOutputFile {
    fn set_frame_buffer(&mut self, pixels: &[Rgba], x_stride: i32, y_stride: i32) -> Result<()> {
        // do some basic bounds checking 
        if pixels.len() < (x_stride * y_stride) { return Err(Error::BufferTooSmall) }

        unsafe {
            // pass the pixels pointer straight to C++. Note that we are NOT telling Rust about the implicit borrow here. This is
            // not itself unsafe and if we never call any other method that tries to use the pointer it is completely safe
            sys::Imf_RgbaOutputFile_setFrameBuffer(self.0, pixels.as_ptr(), x_stride, y_stride).ok()?;
        }

        Ok(())
    }

    // We mark this function unsafe because this is where C++ will try and use the pixels and any badness will actually happen
    unsafe fn write_pixels(&mut self, num_scanlines: usize) -> Result<()> {
        sys::Imf_RgbaOutputFile_writePixels(self.0, num_scanlines).ok()?;
        Ok(())
    }
}
```

Note that the API itself is Rustic, and only the method that will actually try to do something unsafe is marked as such. Thus the resulting API is still ergonomic to use and follows Rust's safety rules in the strictest sense (if not in spirit).

It's also worth noting that this is the stage at which the openexr crate is currently, although not all methods that should be marked unsafe according to this scheme are marked as such (it was as I was planning a fully safe layer over the top that this idea came into focus).

## Proposal

I would like to propose that we explicitly specify that our "named -rs crates" (by this I mean the crates that share the name with the library we're binding and would ultimately like to version in lockstep) would conform to this "unsafe+" API and no more. In particular this means we don't get cute with safe abstractions that may limit functionality or performance compared to the C++ library.

This has a number of advantages:
1) It makes it much easier to decide when a crate is "done" and can be published and versioned appropriately. Deciding when an abstracted API would be "ready" would be much harder and more subjective, and getting that wrong comes with a high price in terms of API breakage.
2) It allows people to iterate on ideas for a safe API on top of this layer instead having to drop down to the -sys crate. This means it's easier for less experienced Rust developers to get started and more pleasant to do so. People could design crates with wonderfully punny names and abstractions that make sense for their specific use case without worrying about breaking the "official" API. There's no reason that we (the WG) can't do this too as it's also the "fun" part. If we or anyone else comes up with a safe API that we think suits all use cases without downsides, there's no reason we couldn't fold that back into "named crate" at any minor version release.
3) Thanks to its more direct translation, it's quite possible we'd be able to auto-generate large parts of the "unsafe+" API with cppmm too, which would speed up the binding effort considerably.

**Comment by scott-wilson** [comment]
I think the unsafe + API makes sense for all the reasons that you gave. Also, one of our initial goals is to create APIs that are as close to the C++ APIs as possible (and only deviate where it makes sense). So, if we take a look at an API and decide that building a safe interface takes more resources than we have to give, then this doesn't seem to go against the goal. As long as we keep the door open for someone to create official safe interfaces for the unsafe+ interfaces, then I'd say everyone wins.

Asides from that, I think that as long as those interfaces are marked as unsafe, and we have documentation explaining what are the safety requirements of the consumer of the API, then I personally have no issues with it.

**Comment by alexxbb** [comment]
Agree with everything what said about. Building an official safe API would be an enormous work with API breaking every other release. 

I also faced the same dilemma while working on bindings to Houdini Engine. I started building a nice and cute API from the beginning just to realize later on that I'm changing the design every other week. I settled on a safe, but very thin API on top of raw bindgen. But Houdini Engine is nowhere near as complex as USD or OpenEXR for example 🙂

**Comment by anderslanglands** [comment]
Thanks for adding your experience Aleksei that's good to know.

**Comment by lgritz** [comment]
I'm writing this from a totally uninformed standpoint, but my 2c anyway:

I hope that these goals are flexible, depending on how easy or difficult it is to Rustify a particular API. It would be tremendously disappointing if the Rust binding efforts for the major packages ended with the unsafe kind in a way that invited 3rd parties to make (potentially multiple, incompatible) safe Rust bindings on top of them, which would be inherently difficult to synchronize because they and the projects are separately administered, developed, and released. For OpenEXR, for example, we've had no end of trouble from the fact that a 3rd party Python binding proliferated that many people assumed we were responsible for, but we were not. I know that for the projects I'm involved with, the hope would be to have one official best-of-class Rust binding coming from the project itself and synchronized with its releases.

Just thinking about the Python bindings for OIIO, we tried when we could to make an approximately 1:1 correspondence between the Python and the C++ API, but not slavishly so -- we deviated when the C++ API was clearly not pythonic. For example, in C++ all the ImageInput methods that read pixels take raw pointers to buffers, whereas the corresponding Python calls simply return a NumPy array containing the pixel data being asked for. I would like to think that the same approach can be done for Rust: correspond as close as is wise, but no closer.

I'm not sure what you mean by "breaking every other release." OpenEXR and OIIO, for example, break API back-compatibility only once every several years, and we're currently batting around ideas to make it so that even ABI does not break annually as it currently does.

Furthermore, if you can identify parts of the C++ APIs that are difficult to turn into safe Rust, then as far as I'm concerned, that means that we should consider changing the C++ (or at least adding something better). I don't think it's a coincidence that your example of OpenEXR RgbaInputFile/setFrameBuffer/writePixels has long been identified as a problematic part of the C++ API and is slated for replacement. (Specifically, we fully understand that setFrameBuffer and writePixels should never have been split into two separate calls, since that creates hidden state that is not thread-safe and requires careful locking by the caller; the new OpenEXRCore API does not have that problem.)

I can't speak for all the projects, but certainly for the ones I'm involved with, I would ask: "Please pressure us to make our APIs more amenable to proper Rust wrapping, rather than make a Rust binding that only takes us half way."

**Comment by anderslanglands** [comment]
Thanks Larry, that's valuable feedback.

To be clear, having a few functions in the crate marked as unsafe isn't
throwing our hands in the air and saying "good luck" - you're still much
better off than in C++ and the API's we're creating are still pretty
idiomatic Rust (even more so than your Python example). Things like getting
an ImfHeader or an ImageSpec reference from an image can still have their
lifetimes tracked easily enough by the compiler so we're always going to do
that. We're limiting those unsafe methods to things that can't be expressed
safely without clear tradeoffs (adding Mutexes etc). I haven't done a full
audit of OpenEXR yet, but I *think* there it would just be the read/write
methods we'd need to mark unsafe.

Moreover, I'm not suggesting that we never try and find a good solution for
these problems, just that we don't set out to try and solve everything
straight away and end up tying ourselves in knots trying to find an ideal
solution. I'd rather get something out that works and is useful and get
some experience using it before going that last mile. In fact, just
grepping the resulting crate for "unsafe" would nicely highlight all the
problem spots in the C++ API that we might want to take upstream to fix.
OpenEXR is a great example of where we can (and will) do that.

With regards assumed support for 3rd party bindings, that's much less of a
problem for Rust. For better or worse, Rust's package repository, crates.io,
is not namespaced, and we (well the LF) now own all the relevant crate
names so someone can't just publish a crate called "openexr" now (although
the problem doesn't *completely* go away--someone could call one 'exr' for
example). In fact I think this proposed scheme actually helps with that
rather than hurts since it's much more likely someone wanting a different
API for working with openexr files would choose to depend on our crate
rather than binding it themselves, *because* we haven't locked the
downstream user into architectural decisions they may not agree with.

What I mean by 'breaking every other release' is that finding the right
safe API is a process that will take a lot of iteration, driven by
real-world feedback. Making changes to that API either means bumping the
major version regularly, which is not something we want to inflict on
upstream (particularly for libraries like OpenEXR and OIIO which are
stable), or staying pre-1.0 until we think it's "done", which could easily
lead to eternal bikeshedding, or just giving up on version locking
altogether, which would be a shame. By contrast, I see this as a route to
getting us quickly to a stable foundation we can build on, while also
effectively highlighting the areas that need attention that we can bring
upstream to the C++ devs, and since it's only those parts of the API that
need changing, it's much easier to get everyone to make breaking changes
together at that point.

**Comment by lgritz** [comment]
Thanks for the clarification. I think I was afraid that you had even more significantly cut back on the scope of what you hoped to accomplish.

**Comment by luke-titley** [comment]
I'm a bit late to the conversation. This is a summary of the points which I understood and agree with.

* thin layer on top of c++ api is a tangible, measurable goal.
* we'll discover the limitations of c++ api (in terms safety) while doing that thin layer.
* it puts the responsibility of making a safe api back on the original library developers (which is where it should be).
* stating that as a goal encourages other rust developers in the community to prototype safer APIs which can feed back into the original c++ libs.
